### PR TITLE
[Feature] Add reduction parameter to On-Policy losses.

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

This PR introduces a reduction option to the on-policy losses, similar to how Torch does it (e.g. https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html).

The ideas is to validate the approach for on-policy losses, the then move on to replicate it in the other losses.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: test/test_cost.py [hunk 1]
```diff
@@ -155,6 +155,10 @@
 )
 
 
+# Increase recursion limit, some objectives have more than 1000 tests
+sys.setrecursionlimit(2000)
```
Comment (albertbou92) [inline test/test_cost.py:159]: I needed to increase this. Otherwise classes with more than 1000 tests fail after hitting this number

File: torchrl/objectives/a2c.py [hunk 2]
```diff
@@ -234,6 +240,7 @@ def __init__(
         functional: bool = True,
         actor: ProbabilisticTensorDictSequential = None,
         critic: ProbabilisticTensorDictSequential = None,
+        reduction: str = "mean",
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:243]: I would use `None` as a default, it can come in handy if some day we change the default
smth like `if reduction is None: reduction="mean"` later in the constructor.

File: torchrl/objectives/a2c.py [hunk 3]
```diff
@@ -388,7 +396,7 @@ def get_entropy_bonus(self, dist: d.Distribution) -> torch.Tensor:
             entropy = dist.entropy()
         except NotImplementedError:
             x = dist.rsample((self.samples_mc_entropy,))
-            entropy = -dist.log_prob(x)
+            entropy = -dist.log_prob(x).mean(0)
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:399]: why 0? What if the input tensordict has 2 or 3 dims? Don't we want to mean across the number of dims of the input tensordict?

File: torchrl/objectives/a2c.py [hunk 4]
```diff
@@ -457,14 +465,16 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
-            td_out.set("loss_entropy", -self.entropy_coef * entropy.mean())
+            td_out.set("entropy", entropy.detach())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
-            loss_critic = self.loss_critic(tensordict).mean()
-            td_out.set("loss_critic", loss_critic.mean())
+            loss_critic = self.loss_critic(tensordict)
+            td_out.set("loss_critic", loss_critic)
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:477]:
```suggestion
            td_out = td_out.apply(functools.partial(_reduce, reduction=self.reduction), batch_size=[])
```

File: torchrl/objectives/ppo.py [hunk 5]
```diff
@@ -85,6 +86,10 @@ class PPOLoss(LossModule):
             Functionalizing permits features like meta-RL, but makes it
             impossible to use distributed models (DDP, FSDP, ...) and comes
             with a little cost. Defaults to ``True``.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+            ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
+            ``"mean"``: the sum of the output will be divided by the number of
+            elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:92]: ditto

File: torchrl/objectives/ppo.py [hunk 6]
```diff
@@ -278,6 +283,7 @@ def __init__(
         functional: bool = True,
         actor: ProbabilisticTensorDictSequential = None,
         critic: ProbabilisticTensorDictSequential = None,
+        reduction: str = "mean",
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:286]: ditto

File: torchrl/objectives/ppo.py [hunk 7]
```diff
@@ -444,7 +451,7 @@ def get_entropy_bonus(self, dist: d.Distribution) -> torch.Tensor:
             entropy = dist.entropy()
         except NotImplementedError:
             x = dist.rsample((self.samples_mc_entropy,))
-            entropy = -dist.log_prob(x)
+            entropy = -dist.log_prob(x).mean(0)
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:454]: ditto

File: torchrl/objectives/ppo.py [hunk 8]
```diff
@@ -530,15 +537,20 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
-            td_out.set("loss_entropy", -self.entropy_coef * entropy.mean())
+            td_out.set("entropy", entropy.detach())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
-            loss_critic = self.loss_critic(tensordict).mean()
-            td_out.set("loss_critic", loss_critic.mean())
+            loss_critic = self.loss_critic(tensordict)
+            td_out.set("loss_critic", loss_critic)
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:552]:
```suggestion
            td_out = td_out.apply(functools.partial(_reduce, reduction=self.reduction), batch_size=[])
```

File: torchrl/objectives/ppo.py [hunk 9]
```diff
@@ -692,6 +708,7 @@ def __init__(
         normalize_advantage: bool = True,
         gamma: float = None,
         separate_losses: bool = False,
+        reduction: str = "mean",
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:711]: ditto

File: torchrl/objectives/ppo.py [hunk 10]
```diff
@@ -634,6 +646,10 @@ class ClipPPOLoss(PPOLoss):
             Functionalizing permits features like meta-RL, but makes it
             impossible to use distributed models (DDP, FSDP, ...) and comes
             with a little cost. Defaults to ``True``.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+            ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
+            ``"mean"``: the sum of the output will be divided by the number of
+            elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:652]: ditto

File: torchrl/objectives/ppo.py [hunk 11]
```diff
@@ -764,16 +782,19 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
+            td_out.set("entropy", entropy.detach())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
             loss_critic = self.loss_critic(tensordict)
-            td_out.set("loss_critic", loss_critic.mean())
-        td_out.set("ESS", ess.mean() / batch)
+            td_out.set("loss_critic", loss_critic)
+
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:796]:
```suggestion
            td_out = td_out.apply(functools.partial(_reduce, reduction=self.reduction), batch_size=[])
```

File: torchrl/objectives/ppo.py [hunk 12]
```diff
@@ -764,16 +782,19 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
+            td_out.set("ESS", _reduce(ess, self.reduction) / batch)
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:797]: it's a bit problematic that the output depends on the reduction... 
I'm not sure that having the batch-size set is going to work across all losses anyway (some losses subsample a part of data to compute part of the loss).
I agree it isn't super clear though but maybe the best we can do to keep things generic.

File: torchrl/objectives/ppo.py [hunk 13]
```diff
@@ -832,7 +853,10 @@ class KLPENPPOLoss(PPOLoss):
             Functionalizing permits features like meta-RL, but makes it
             impossible to use distributed models (DDP, FSDP, ...) and comes
             with a little cost. Defaults to ``True``.
-
+        reduction (str, optional): Specifies the reduction to apply to the output:
+            ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
+            ``"mean"``: the sum of the output will be divided by the number of
+            elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:859]: ditto

File: torchrl/objectives/ppo.py [hunk 14]
```diff
@@ -895,6 +919,7 @@ def __init__(
         normalize_advantage: bool = True,
         gamma: float = None,
         separate_losses: bool = False,
+        reduction: str = "mean",
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:922]: ditto

File: torchrl/objectives/ppo.py [hunk 15]
```diff
@@ -978,20 +1004,21 @@ def forward(self, tensordict: TensorDictBase) -> TensorDict:
...
-            td_out.set("loss_entropy", -self.entropy_coef * entropy.mean())
-
+            td_out.set("entropy", entropy.detach())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
             loss_critic = self.loss_critic(tensordict)
-            td_out.set("loss_critic", loss_critic.mean())
+            td_out.set("loss_critic", loss_critic)
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:1021]:
```suggestion
            td_out = td_out.apply(functools.partial(_reduce, reduction=self.reduction), batch_size=[])
```

File: torchrl/objectives/reinforce.py [hunk 16]
```diff
@@ -60,6 +62,10 @@ class ReinforceLoss(LossModule):
             Functionalizing permits features like meta-RL, but makes it
             impossible to use distributed models (DDP, FSDP, ...) and comes
             with a little cost. Defaults to ``True``.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+            ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
+            ``"mean"``: the sum of the output will be divided by the number of
+            elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
```
Comment (vmoens) [inline torchrl/objectives/reinforce.py:68]: ditto

File: torchrl/objectives/reinforce.py [hunk 17]
```diff
@@ -388,10 +396,13 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
-        loss_actor = loss_actor.mean()
-        td_out = TensorDict({"loss_actor": loss_actor}, [])
+        td_out = TensorDict(
+            {"loss_actor": loss_actor}, batch_size=tensordict.batch_size
+        )
 
-        td_out.set("loss_value", self.loss_critic(tensordict).mean())
+        td_out.set("loss_value", self.loss_critic(tensordict))
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
```
Comment (vmoens) [inline torchrl/objectives/reinforce.py:405]:
```suggestion
            td_out = td_out.apply(functools.partial(_reduce, reduction=self.reduction), batch_size=[])
```

File: torchrl/objectives/reinforce.py [hunk 18]
```diff
@@ -223,6 +229,7 @@ def __init__(
         functional: bool = True,
         actor: ProbabilisticTensorDictSequential = None,
         critic: ProbabilisticTensorDictSequential = None,
+        reduction: str = "mean",
```
Comment (vmoens) [inline torchrl/objectives/reinforce.py:232]: ditto

**Review by vmoens — COMMENTED** [review]
Great stuff!
I wonder if long term we should not structure the loss better, to avoid the batch-size issue pointed here.

```python
TensorDict({"loss": {"actor": tensor, ...}, "metadata": {...}}, [])
```
which could let us set a different batch-size at diferent levels.
It isn't going to be easy to move to that format though! So for now I think the best would be to keep the output without batch-size until we figure out how to account for it long-term.
```

File: torchrl/objectives/a2c.py (hunk 3 above)
Comment (albertbou92) [reply torchrl/objectives/a2c.py:399]: Because dist.rsample generates `self.samples_mc_entropy` samples along dimension 0. So we average then also in dim 0 to get an estimation of the entropy for each position of the input tensordict.

File: torchrl/objectives/a2c.py (hunk 2 above)
Comment (albertbou92) [reply torchrl/objectives/a2c.py:243]: I am not sure if I follow. Default `mean` ensures BC and seems to me a a reasonable choice. Also Torch losses have `mean` as default, which I think is consistent.

File: torchrl/objectives/a2c.py [hunk 19]
```diff
@@ -68,6 +70,10 @@ class A2CLoss(LossModule):
             Functionalizing permits features like meta-RL, but makes it
             impossible to use distributed models (DDP, FSDP, ...) and comes
             with a little cost. Defaults to ``True``.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+            ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
+            ``"mean"``: the sum of the output will be divided by the number of
+            elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply torchrl/objectives/a2c.py:76]: I used `str`'s because I was trying to be consistent with torch losses, which don't use Enum. I am fine using an Enum if preferred.

File: torchrl/objectives/ppo.py (hunk 12 above)
Comment (albertbou92) [reply torchrl/objectives/ppo.py:797]: What about adding all losses with the correct shape in the first level and add a nested TensorDict with all the other tensors. We can call it `metadata` as suggested in the first comment.

**Comment by albertbou92** [comment]
I incorporated a part of the suggestions and left comments in the points that might need further discussion!

File: torchrl/objectives/a2c.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/objectives/a2c.py:77]: Let's use strings for now. I like enum but it's also always the one more import that doens't really make things that much easier...

File: torchrl/objectives/a2c.py (hunk 2 above)
Comment (vmoens) [reply torchrl/objectives/a2c.py:244]: It will still be "mean". But not having "mean" is good practice because it allows us to detect whenever the user is passing or not the reduction. 
Might come in handy at some point: for instance, if we want to change the default for one loss someday, we will just have to check whether the reduction is None or not.

File: torchrl/objectives/a2c.py (hunk 3 above)
Comment (vmoens) [reply torchrl/objectives/a2c.py:400]: Oh right!

File: torchrl/objectives/ppo.py (hunk 12 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:808]: that won't work because the nested tensordict must have the same leading dims as the root (+ other optional dims) as it is the case for tensors

File: torchrl/objectives/a2c.py (hunk 2 above)
Comment (albertbou92) [reply torchrl/objectives/a2c.py:244]: and how do you define a no-reduction option then?

File: torchrl/objectives/ppo.py (hunk 12 above)
Comment (albertbou92) [reply torchrl/objectives/ppo.py:808]: aah true...

File: torchrl/objectives/a2c.py (hunk 2 above)
Comment (vmoens) [reply torchrl/objectives/a2c.py:244]: `"none"` (not `None`)

File: test/test_cost.py [hunk 20]
```diff
@@ -7144,6 +7184,9 @@ def test_reinforce_value_net(
             elif td_est is not None:
                 loss_fn.make_value_estimator(td_est)
             loss_td = loss_fn(td)
+            if reduction == "none":
+                assert loss_td.batch_size == td.batch_size
+                loss_td = loss_td.apply(lambda x: x.float().mean(), batch_size=[])
```
Comment (vmoens) [inline test/test_cost.py:7189]: maybe here instead of converting to float we can do
```python
def func(x):
    if x.dtype != torch.float:
        return
    return x.mean()
loss_td = loss_td.apply(func, batch_size=[])
```
because I suspect we convert to float things that we may not want to convert to float (eg metadata)

**Review by vmoens — APPROVED** [review]
LGTM just smth minor in the tests
